sys/shell/cmds/Kconfig

```diff
@@ -0,0 +1,284 @@
+# Copyright (c) 2022 Otto-von-Guericke-Universität Magdeburg
+#
+# This file is subject to the terms and conditions of the GNU Lesser
+# General Public License v2.1. See the file LICENSE in the top level
+# directory for more details.
+#
+
+menuconfig MODULE_SHELL_CMDS
+    bool "Support for shell commands"
+    depends on MODULE_SHELL
+
+config MODULE_SHELL_COMMANDS
+    bool "Default shell commands for enabled modules"
+    select MODULE_SHELL_CMDS
+    select MODULE_SHELL_CMD_APP_METADATA if MODULE_APP_METADATA
+    select MODULE_SHELL_CMD_AT30TSE75X if MODULE_AT30TSE75X
```

Comment:
This is apparently considered as `MODULE_AT30TSE75X` depending on `MODULE_SHELL_CMD_AT30TSE75X`, resulting in a dependency loop.

If I go for

```Kconfig
config MODULE_SHELL_CMD_AT30TSE75X
    bool "Command to testAT30TSE75x temperature sensors"
    default MODULE_SHELL_COMMANDS && MODULE_AT30TSE75X
    depends on MODULE_SHELL_CMDS
    select MODULE_AT30TSE75X
```

instead (which seems to also be cleaner anyway to me), I again have a dependency loop.

What would be the correct way to model:

- `MODULE_SHELL_CMD_AT30TSE75X` pulls in `MODULE_AT30TSE75X`
- `MODULE_SHELL_CMD_AT30TSE75X` depends on `MODULE_SHELL_CMDS`
- `MODULE_SHELL_COMMANDS` pulls in `MODULE_SHELL_CMDS`
- if `MODULE_SHELL_COMMANDS` *and* `MODULE_AT30TSE75X` is already selected, also select `MODULE_SHELL_CMD_AT30TSE75X`
 
?

Maybe @leandrolanzieri knows?

Reply:
The problem is this `MODULE_SHELL_CMD_AT30TSE75X` selecting `MODULE_AT30TSE75X` and this idea of is already selected.

The first and simplest solution if you want to keep the current design pattern is having a seperate symbol that brings in the module:

```kconfig
config FORCE_MODULE_SHELL_CMD_AT30TSE75X
    bool "description for prompt"
    select MODULE_AT30TSE75X
    select FORCE_MODULE_SHELL_CMD_AT30TSE75X

config MODULE_SHELL_CMD_AT30TSE75X
    bool
    depends on MODULE_AT30TSE75X
```

See below for more examples.

The other way can be closer to `saul_default` where the logic in not in the shell commands but in each module that is providing the cmds.

<details><summary>half finished diff of suggestion</summary>

```diff
diff --git a/sys/shell/cmds/Kconfig b/sys/shell/cmds/Kconfig
index a7189f218d..3fe7888da7 100644
--- a/sys/shell/cmds/Kconfig
+++ b/sys/shell/cmds/Kconfig
@@ -5,198 +5,213 @@
 # directory for more details.
 #
 
-menuconfig MODULE_SHELL_CMDS
-    bool "Support for shell commands"
-    depends on MODULE_SHELL
-
 config MODULE_SHELL_COMMANDS
     bool "Default shell commands for enabled modules"
+    depends on MODULE_SHELL
     select MODULE_SHELL_CMDS
     select MODULE_SHELL_CMD_APP_METADATA if MODULE_APP_METADATA
     select MODULE_SHELL_CMD_AT30TSE75X if MODULE_AT30TSE75X
-    select MODULE_SHELL_CMD_BENCHMARK_UDP if MODULE_BENCHMARK_UDP
-    select MODULE_SHELL_CMD_CCN-LITE-UTILS if MODULE_CCN-LITE-UTILS
-    select MODULE_SHELL_CMD_CONN_CAN if MODULE_CONN_CAN
-    select MODULE_SHELL_CMD_CORD_EP if MODULE_CORD_EP
-    select MODULE_SHELL_CMD_CRYPTOAUTHLIB if MODULE_CRYPTOAUTHLIB
-    select MODULE_SHELL_CMD_DFPLAYER if MODULE_DFPLAYER
-    select MODULE_SHELL_CMD_FIB if MODULE_FIB
-    select MODULE_SHELL_CMD_GNRC_ICMPV6_ECHO if MODULE_GNRC_ICMPV6_ECHO
-    select MODULE_SHELL_CMD_GNRC_IPV6_BLACKLIST if MODULE_GNRC_IPV6_BLACKLIST
-    select MODULE_SHELL_CMD_GNRC_IPV6_FRAG_STATS if MODULE_GNRC_IPV6_FRAG_STATS
-    select MODULE_SHELL_CMD_GNRC_IPV6_NIB if MODULE_GNRC_IPV6_NIB
-    select MODULE_SHELL_CMD_GNRC_IPV6_WHITELIST if MODULE_GNRC_IPV6_WHITELIST
-    select MODULE_SHELL_CMD_GNRC_NETIF if MODULE_GNRC_NETIF
-    select MODULE_SHELL_CMD_GNRC_PKTBUF if MODULE_GNRC_PKTBUF
-    select MODULE_SHELL_CMD_GNRC_RPL if MODULE_GNRC_RPL
-    select MODULE_SHELL_CMD_GNRC_SIXLOWPAN_CTX if MODULE_GNRC_SIXLOWPAN_CTX
-    select MODULE_SHELL_CMD_GNRC_SIXLOWPAN_FRAG_STATS if MODULE_GNRC_SIXLOWPAN_FRAG_STATS
-    select MODULE_SHELL_CMD_GNRC_UDP if MODULE_GNRC_UDP
-    select MODULE_SHELL_CMD_LWIP_NETIF if MODULE_LWIP_NETIF
-    select MODULE_SHELL_CMD_MCI if MODULE_MCI
-    select MODULE_SHELL_CMD_NANOCOAP_VFS if MODULE_NANOCOAP_VFS
-    select MODULE_SHELL_CMD_NETSTATS_NEIGHBOR if MODULE_NETSTATS_NEIGHBOR
-    select MODULE_SHELL_CMD_NIMBLE_NETIF if MODULE_NIMBLE_NETIF
-    select MODULE_SHELL_CMD_NIMBLE_STATCONN if MODULE_NIMBLE_STATCONN
-    select MODULE_SHELL_CMD_OPENWSN if MODULE_OPENWSN
-    select MODULE_SHELL_CMD_PS if MODULE_PS
-    select MODULE_SHELL_CMD_PM if HAS_PERIPH_PM
-    select MODULE_SHELL_CMD_RTC if HAS_PERIPH_RTC || MODULE_RTT_RTC
-    select MODULE_SHELL_CMD_HEAP if MODULE_HEAP_CMD || MODULE_LPC2387
-    select MODULE_SHELL_CMD_I2C_SCAN if MODULE_I2C_SCAN
-    select MODULE_SHELL_CMD_MD5SUM if MODULE_MD5SUM
-    select MODULE_SHELL_CMD_NICE if MODULE_NICE
-    select MODULE_SHELL_CMD_RANDOM if MODULE_RANDOM_CMD
+    # select MODULE_SHELL_CMD_BENCHMARK_UDP if MODULE_BENCHMARK_UDP
+    # select MODULE_SHELL_CMD_CCN-LITE-UTILS if MODULE_CCN-LITE-UTILS
+    # select MODULE_SHELL_CMD_CONN_CAN if MODULE_CONN_CAN
+    # select MODULE_SHELL_CMD_CORD_EP if MODULE_CORD_EP
+    # select MODULE_SHELL_CMD_CRYPTOAUTHLIB if MODULE_CRYPTOAUTHLIB
+    # select MODULE_SHELL_CMD_DFPLAYER if MODULE_DFPLAYER
+    # select MODULE_SHELL_CMD_FIB if MODULE_FIB
+    # select MODULE_SHELL_CMD_GNRC_ICMPV6_ECHO if MODULE_GNRC_ICMPV6_ECHO
+    # select MODULE_SHELL_CMD_GNRC_IPV6_BLACKLIST if MODULE_GNRC_IPV6_BLACKLIST
+    # select MODULE_SHELL_CMD_GNRC_IPV6_FRAG_STATS if MODULE_GNRC_IPV6_FRAG_STATS
+    # select MODULE_SHELL_CMD_GNRC_IPV6_NIB if MODULE_GNRC_IPV6_NIB
+    # select MODULE_SHELL_CMD_GNRC_IPV6_WHITELIST if MODULE_GNRC_IPV6_WHITELIST
+    # select MODULE_SHELL_CMD_GNRC_NETIF if MODULE_GNRC_NETIF
+    # select MODULE_SHELL_CMD_GNRC_PKTBUF if MODULE_GNRC_PKTBUF
+    # select MODULE_SHELL_CMD_GNRC_RPL if MODULE_GNRC_RPL
+    # select MODULE_SHELL_CMD_GNRC_SIXLOWPAN_CTX if MODULE_GNRC_SIXLOWPAN_CTX
+    # select MODULE_SHELL_CMD_GNRC_SIXLOWPAN_FRAG_STATS if MODULE_GNRC_SIXLOWPAN_FRAG_STATS
+    # select MODULE_SHELL_CMD_GNRC_UDP if MODULE_GNRC_UDP
+    # select MODULE_SHELL_CMD_LWIP_NETIF if MODULE_LWIP_NETIF
+    # select MODULE_SHELL_CMD_MCI if MODULE_MCI
+    # select MODULE_SHELL_CMD_NANOCOAP_VFS if MODULE_NANOCOAP_VFS
+    # select MODULE_SHELL_CMD_NETSTATS_NEIGHBOR if MODULE_NETSTATS_NEIGHBOR
+    # select MODULE_SHELL_CMD_NIMBLE_NETIF if MODULE_NIMBLE_NETIF
+    # select MODULE_SHELL_CMD_NIMBLE_STATCONN if MODULE_NIMBLE_STATCONN
+    # select MODULE_SHELL_CMD_OPENWSN if MODULE_OPENWSN
+    # select MODULE_SHELL_CMD_PS if MODULE_PS
+    # select MODULE_SHELL_CMD_PM if HAS_PERIPH_PM
+    # select MODULE_SHELL_CMD_RTC if HAS_PERIPH_RTC || MODULE_RTT_RTC
+    # select MODULE_SHELL_CMD_HEAP if MODULE_HEAP_CMD || MODULE_LPC2387
+    # select MODULE_SHELL_CMD_I2C_SCAN if MODULE_I2C_SCAN
+    # select MODULE_SHELL_CMD_MD5SUM if MODULE_MD5SUM
+    # select MODULE_SHELL_CMD_NICE if MODULE_NICE
+    # select MODULE_SHELL_CMD_RANDOM if MODULE_RANDOM_CMD
 
     help
         Shell commands can be enabled and disabled individually as dedicated
         modules. This module instead will select the shell commands of modules
         already used, if this is considered as "good default choice".
 
-config MODULE_SHELL_CMD_APP_METADATA
+
+menuconfig MODULE_SHELL_CMDS
+    bool "Support for shell commands"
+    depends on MODULE_SHELL
+
+if MODULE_SHELL_CMDS
+
+config FORCE_MODULE_SHELL_CMD_APP_METADATA
     bool "Command to print app meta data as JSON"
-    depends on MODULE_SHELL_CMDS
     select MODULE_APP_METADATA
+    select MODULE_SHELL_CMD_APP_METADATA
 
-config MODULE_SHELL_CMD_AT30TSE75X
-    bool "Command to testAT30TSE75x temperature sensors"
-    depends on MODULE_SHELL_CMDS
+config MODULE_SHELL_CMD_APP_METADATA
+    bool
+    depends on MODULE_APP_METADATA
+
+config FORCE_MODULE_SHELL_CMD_AT30TSE75X
+    bool "Command to test AT30TSE75x temperature sensors"
     select MODULE_AT30TSE75X
+    select MODULE_SHELL_CMD_AT30TSE75X
 
-config MODULE_SHELL_CMD_BENCHMARK_UDP
+config MODULE_SHELL_CMD_AT30TSE75X
+    bool
+    depends on MODULE_AT30TSE75X
+
+config FORCE_MODULE_SHELL_CMD_BENCHMARK_UDP
     bool "Command to perform a UDP benchmark"
-    depends on MODULE_SHELL_CMDS
     select MODULE_BENCHMARK_UDP
+    select MODULE_SHELL_CMD_BENCHMARK_UDP
+
+config MODULE_SHELL_CMD_BENCHMARK_UDP
+    bool
+    depends on MODULE_BENCHMARK_UDP
 
-config MODULE_SHELL_CMD_CCN-LITE-UTILS
+config FORCE_MODULE_SHELL_CMD_CCN_LITE_UTILS
     bool "Commands to interact with the CCN-Lite stack"
-    depends on MODULE_SHELL_CMDS
-    select MODULE_CCN-LITE-UTILS
+    select MODULE_CCN_LITE_UTILS
+    select MODULE_SHELL_CMD_CCN_LITE_UTILS
 
-config MODULE_SHELL_CMD_CONN_CAN
+config MODULE_SHELL_CMD_CCN_LITE_UTILS
+    bool
+    depends on MODULE_CCN_LITE_UTILS
+
+config FORCE_MODULE_SHELL_CMD_CONN_CAN
     bool "Command to interact with the CAN stack"
-    depends on MODULE_SHELL_CMDS
     select MODULE_CONN_CAN
+    select MODULE_SHELL_CMD_CONN_CAN
 
-config MODULE_SHELL_CMD_CORD_EP
+config MODULE_SHELL_CMD_CONN_CAN
+    bool
+    depends on MODULE_CONN_CAN
+
+config FORCE_MODULE_SHELL_CMD_CORD_EP
     bool "Command to interact with a resource directory endpoint"
-    depends on MODULE_SHELL_CMDS
     select MODULE_CORD_EP
+    select MODULE_SHELL_CMD_CORD_EP
 
-config MODULE_SHELL_CMD_CRYPTOAUTHLIB
+config MODULE_SHELL_CMD_CORD_EP
+    bool
+    depends on MODULE_CORD_EP
+
+config FORCE_MODULE_SHELL_CMD_CRYPTOAUTHLIB
     bool "Command to interact with Microchip CryptoAuth devices"
-    depends on MODULE_SHELL_CMDS
     select MODULE_CRYPTOAUTHLIB
+    select MODULE_SHELL_CMD_CRYPTOAUTHLIB
 
-config MODULE_SHELL_CMD_DFPLAYER
+config MODULE_SHELL_CMD_CRYPTOAUTHLIB
+    bool
+    depends on MODULE_CRYPTOAUTHLIB
+
+config FORCE_MODULE_SHELL_CMD_DFPLAYER
     bool "Command to control a dfplayer MP3 player"
-    depends on MODULE_SHELL_CMDS
     select MODULE_DFPLAYER
+    select MODULE_SHELL_CMD_DFPLAYER
+
+config MODULE_SHELL_CMD_DFPLAYER
+    bool
+    depends on MODULE_DFPLAYER
 
 config MODULE_SHELL_CMD_FIB
     bool "Command to interact with the Forwarding Information Base (FIB)"
-    depends on MODULE_SHELL_CMDS
     select MODULE_FIB
     select MODULE_POSIX_INET
 
 config MODULE_SHELL_CMD_GNRC_ICMPV6_ECHO
     bool "Command to send ICMPv6 Echo Requests (a.k.a. ping)"
-    depends on MODULE_SHELL_CMDS
     select MODULE_GNRC_ICMPV6_ECHO
     select MODULE_NETUTILS
     select MODULE_ZTIMER_USEC
 
 config MODULE_SHELL_CMD_GNRC_IPV6_BLACKLIST
     bool "Command to manage IPv6 addresses in reception deny list"
-    depends on MODULE_SHELL_CMDS
     select MODULE_GNRC_IPV6_BLACKLIST
 
 config MODULE_SHELL_CMD_GNRC_IPV6_FRAG_STATS
     bool "Command to show IPv6 fragmentation statistics"
-    depends on MODULE_SHELL_CMDS
     select MODULE_GNRC_IPV6_FRAG_STATS
 
 config MODULE_SHELL_CMD_GNRC_IPV6_NIB
     bool ""
-    depends on MODULE_SHELL_CMDS
 
 config MODULE_SHELL_CMD_GNRC_IPV6_WHITELIST
     bool "Command to manage IPv6 addresses in reception allow list"
-    depends on MODULE_SHELL_CMDS
     select MODULE_GNRC_IPV6_WHITELIST
 
 config MODULE_SHELL_CMD_GNRC_NETIF
     bool "Command to manage GNRC network interfaces (ifconfig)"
-    depends on MODULE_SHELL_CMDS
     select MODULE_GNRC_NETIF
 
 config MODULE_SHELL_CMD_GNRC_PKTBUF
     bool "Command to print stats of the GNRC packet buffer"
-    depends on MODULE_SHELL_CMDS
     select MODULE_GNRC_PKTBUF
 
 config MODULE_SHELL_CMD_GNRC_RPL
     bool "Command to configure GNRC's RPL implementation"
-    depends on MODULE_SHELL_CMDS
     select MODULE_GNRC_RPL
 
 config MODULE_SHELL_CMD_GNRC_SIXLOWPAN_CTX
     bool "Command to configure 6LoWPAN context in GNRC"
-    depends on MODULE_SHELL_CMDS
     select MODULE_SIXLOWPAN_CTX
 
 config MODULE_SHELL_CMD_GNRC_SIXLOWPAN_FRAG_STATS
     bool "Command to display 6LoWPAN fragment statistics"
-    depends on MODULE_SHELL_CMDS
     select MODULE_GNRC_SIXLOWPAN_FRAG_STATS
 
 config MODULE_SHELL_CMD_GNRC_UDP
     bool "Command interface to a UDP server and client"
-    depends on MODULE_SHELL_CMDS
     select MODULE_GNRC_UDP
     select MODULE_GNRC_PKTDUMP
 
 config MODULE_SHELL_CMD_HEAP
     bool "Command to print dynamic memory allocation statistics"
-    depends on MODULE_SHELL_CMDS
 
 config MODULE_SHELL_CMD_I2C_SCAN
     bool "Command to scan for I2C devices"
-    depends on MODULE_SHELL_CMDS
     depends on MODULE_PERIPH_I2C
 
 config MODULE_SHELL_CMD_LWIP_NETIF
     bool "Command to manage lwIP network interfaces (ifconfig)"
-    depends on MODULE_SHELL_CMDS
     select MODULE_LWIP_NETIF
 
 config MODULE_SHELL_CMD_MCI
     bool "Commands to query parameters and read contents from memory cards"
-    depends on MODULE_SHELL_CMDS
     select MODULE_MCI
 
 config MODULE_SHELL_CMD_MD5SUM
     bool "Command to calculate the MD5 sum of a file"
-    depends on MODULE_SHELL_CMDS
     select MODULE_SHELL_CMD_VFS
 
 config MODULE_SHELL_CMD_NANOCOAP_VFS
     bool "Commands to upload/download files to/from a CoAP server"
-    depends on MODULE_SHELL_CMDS
     select MODULE_NANOCOAP_VFS
 
 config MODULE_SHELL_CMD_NETSTATS_NEIGHBOR
     bool "Command to show neighbor statistics"
-    depends on MODULE_SHELL_CMDS
     select MODULE_NETSTATS_NEIGHBOR
 
 config MODULE_SHELL_CMD_NICE
     bool "Command to change the priority of running threads"
-    depends on MODULE_SHELL_CMDS
 
 config MODULE_SHELL_CMD_NIMBLE_NETIF
     bool "Command to manage BLE connections for NimBLE"
-    depends on MODULE_SHELL_CMDS
     select MODULE_FMT
     select MODULE_NIMBLE_NETIF
     select MODULE_NIMBLE_SCANLIST
@@ -204,81 +219,67 @@ config MODULE_SHELL_CMD_NIMBLE_NETIF
 
 config MODULE_SHELL_CMD_NIMBLE_STATCONN
     bool "Command to controle the NimBLE netif statconn connection manager"
-    depends on MODULE_SHELL_CMDS
     select MODULE_NIMBLE_STATCONN
 
 config MODULE_SHELL_CMD_OPENWSN
     bool "Commands to interact with the OpenWSN network statck (ifconfig, openwsn)"
-    depends on MODULE_SHELL_CMDS
     select MODULE_OPENWSN
     select MODULE_NETIF
     select MODULE_IPV6_ADDR
 
 config MODULE_SHELL_CMD_PM
     bool "Command to interact with the layered power management subsystem"
-    depends on MODULE_SHELL_CMDS
     depends on HAS_PERIPH_PM
 
 config MODULE_SHELL_CMD_PS
     bool "Command to print information about running threads"
-    depends on MODULE_SHELL_CMDS
     select MODULE_PS
 
 config MODULE_SHELL_CMD_RANDOM
     bool "Commands to initialize the PRNG and print 32 bit pseudo-random numbers"
-    depends on MODULE_SHELL_CMDS
     select MODULE_RANDOM
 
 config MODULE_SHELL_CMD_RTC
     bool "Command to control the peripheral real time clock"
-    depends on MODULE_SHELL_CMDS
     depends on HAS_PERIPH_RTC || HAS_PERIPH_RTT
     select MODULE_RTT_RTC if !HAS_PERIPH_RTC
 
 config MODULE_SHELL_CMD_RTT
     bool "Command to control the peripheral real time timer"
-    depends on MODULE_SHELL_CMDS
     depends on HAS_PERIPH_RTT
 
 config MODULE_SHELL_CMD_SAUL_REG
     bool "Command to read sensors and control actuators via SAUL"
-    depends on MODULE_SHELL_CMDS
     select MODULE_SAUL_REG
 
 config MODULE_SHELL_CMD_SEMTECH-LORAMAC
     bool "Command to control the Semtech LoRaMAC stack"
-    depends on MODULE_SHELL_CMDS
     select MODULE_SEMTECH-LORAMAC
 
 config MODULE_SHELL_CMD_SHA1SUM
     bool "Command to compute the SHA1 sum of a file"
-    depends on MODULE_SHELL_CMDS
     select MODULE_SHELL_CMD_VFS
 
 config MODULE_SHELL_CMD_SHA256SUM
     bool "Command to compute the SHA256 sum of a file"
-    depends on MODULE_SHELL_CMDS
     select MODULE_SHELL_CMD_VFS
 
 config MODULE_SHELL_CMD_SHT1X
     bool "Commands to interact with SHT1x sensors"
-    depends on MODULE_SHELL_CMDS
     select MODULE_SHT1X
 
 config MODULE_SHELL_CMD_SNTP
     bool "Command to synchronize time with an SNTP server"
-    depends on MODULE_SHELL_CMDS
     select MODULE_SNTP
 
 config MODULE_SHELL_CMD_SUIT
     bool "Command to trigger a SUIT firmware update"
-    depends on MODULE_SHELL_CMDS
     select MODULE_SUIT_TRANSPORT_WORKER
 
 config MODULE_SHELL_CMD_SYS
     bool "Common utility commands (version, reboot, bootloader)"
-    depends on MODULE_SHELL_CMDS
 
 config MODULE_SHELL_CMD_VFS
     bool "Commands for the VFS module (ls, vfs)"
-    depends on MODULE_SHELL_CMDS
+
+endif # MODULE_SHELL_CMDS
```
</details>